libs/@local/harpc/tower/src/body/stream.rs

```diff
@@ -0,0 +1,227 @@
+use core::task::ready;
+use std::{
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use bytes::Buf;
+use futures::Stream;
+
+use super::{Body, BodyState, Frame};
+
+pin_project_lite::pin_project! {
+    #[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+    pub struct StreamBody<S> {
+        #[pin]
+        stream: S,
+        is_complete: bool,
+    }
+}
+
+impl<S> StreamBody<S> {
+    pub fn new(stream: S) -> Self {
+        Self {
+            stream,
+            is_complete: false,
+        }
+    }
+}
+
+impl<S, D, C, E> Body for StreamBody<S>
+where
+    S: Stream<Item = Result<Frame<D, C>, E>>,
+    D: Buf,
+{
+    type Control = C;
+    type Data = D;
+    type Error = E;
+
+    fn poll_frame(
+        self: Pin<&mut Self>,
+        cx: &mut Context,
+    ) -> Poll<Option<super::BodyFrameResult<Self>>> {
+        let this = self.project();
+
+        let value = ready!(this.stream.poll_next(cx));
+
+        if value.is_none() {
+            *this.is_complete = true;
+        }
+
+        Poll::Ready(value)
+    }
+
+    fn state(&self) -> Option<BodyState> {
+        self.is_complete.then_some(BodyState::Complete)
+    }
+}
+
+pin_project_lite::pin_project! {
+    #[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+    pub struct BodyStream<B> {
+        #[pin]
+        inner: B,
+    }
+}
+
+impl<B> BodyStream<B> {
+    pub fn new(inner: B) -> Self {
+        Self { inner }
+    }
+
+    pub fn into_data_stream(self) -> BodyDataStream<B> {
+        BodyDataStream { inner: self.inner }
+    }
+
+    pub fn into_control_stream(self) -> BodyControlStream<B> {
+        BodyControlStream { inner: self.inner }
+    }
+}
+
+impl<B> Stream for BodyStream<B>
+where
+    B: Body,
+{
+    type Item = Result<Frame<B::Data, B::Control>, B::Error>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        return self.project().inner.poll_frame(cx);
+    }
+}
+
+pin_project_lite::pin_project! {
+    #[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+    pub struct BodyDataStream<B> {
+        #[pin]
+        inner: B,
+    }
+}
+
+impl<B> Stream for BodyDataStream<B>
+where
+    B: Body,
+{
+    type Item = Result<B::Data, B::Error>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        loop {
+            return match ready!(self.as_mut().project().inner.poll_frame(cx)) {
+                Some(Ok(frame)) => match frame.into_data() {
+                    Ok(data) => Poll::Ready(Some(Ok(data))),
+                    Err(_) => continue,
```

Comment:
Is there a way to log or return the error?

Reply:
The error here is just the frame back, so you don't lose any data, it's more or less a filter.
(Modeled after http)